README.md
<p align="center">
  <img width="350" height="350" src="https://raw.githubusercontent.com/sys113/negar/master/negar.png">
</p>

---
<div align="center">
  
![](https://img.shields.io/github/stars/SYS113/negar.svg)
![](https://img.shields.io/badge/language-python-orange.svg)
![](https://img.shields.io/github/forks/SYS113/negar.svg)
![](https://img.shields.io/github/release/SYS113/negar.svg)
![](https://img.shields.io/github/issues/SYS113/negar.svg)
![](https://img.shields.io/badge/license-MIT-informational.svg)
</div>

---
## *review*
<p align="center">
  <img src="https://raw.githubusercontent.com/sys113/negar/master/example/review.png">
</p>

---
## *what is <ins>negar</ins>*
*call <ins>negar</ins> module in your python source code and write error log & text log to a file</ins> ...<br />*

---
## *installation*

+ #### *install by pip*

      # linux
      
      sudo python3 -m pip install negar
      
      # windows
      
      py -m pip install negar
      
+ #### *install by setup.py*

      # linux
      
      sudo python3 -m pip install https://github.com/sys113/negar/archive/0.6.0.zip
      
      # windows
      
      py -m pip install https://github.com/sys113/negar/archive/0.6.0.zip

---
## *example*
+ #### text function &nbsp;&nbsp;&nbsp;&nbsp;*[example one](https://raw.githubusercontent.com/sys113/negar/master/example/text/method%20one.png)* | *[example two](https://raw.githubusercontent.com/sys113/negar/master/example/text/method%20two.png)*
+ #### error function   &nbsp;&nbsp;*[example one](https://raw.githubusercontent.com/sys113/negar/master/example/error/method%20one.png)* | *[example two](https://raw.githubusercontent.com/sys113/negar/master/example/error/method%20two.png)*

---
## *description*
  + #### *log_text*
    ```python
      # write 'log_text' to log file
      log_text = 'negar' 
    ```
  + #### *line*
    ```python
      # set 'Flase' and 'True' value for show line python file in log file , default is False ... 
      line = True
    ```

  + #### *date*
    ```python
      # set 'Flase' and 'True' value for show date in log file , default is True ... 
      line = True
    ```

  + #### *time*
    ```python
      # set 'Flase' and 'True' value for time in log file , default is True ... 
      line = True
    ```
    
  + #### *save*
    ```python
      # write log to 'negar-log.txt'
      save = 'negar-log.txt' 
    ```
  + #### *size*
    ```python
      # set 'size' for log file 
      size = 2
    ```
  + #### *title*
    ```python
      # set 'title' for write custom log file title , defult is 'city < country < continent | user name | os > os version > architecture'
      title = 'log file title'
    ```
---

## *tips*
+ *log file default size is 2 ...*
+ *log file size range is 1 ... 5 number ...*
+ *maximum size of python file name support is 15 character ...*
+ *maximum number to numbering lines support is 9999999 ...*
+ *maximum python source code line number support is 999999 ...*
+ *default log file name is log.txt ...*
+ *previously defined log file size can'not be resized!<br />*
---
## *copyright*
*copyright <ins>SYS113</ins> - <ins>2019</ins>.*

---
## *license* 
*<ins>MIT</ins> license , please see <ins>LICENSE</ins> file.*

---
## *donate* 
+ *for <ins>iranian</ins> users &nbsp; :  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; <ins>  id pay </ins> &nbsp;&nbsp;&nbsp;&nbsp; - &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; https://idpay.ir/sys113*
+ *for <ins>global</ins> users &nbsp; : &nbsp;<ins>BTC wallet id</ins>&nbsp; - &nbsp; 149JgUmFqG6MvFg79Ldrvdk2bN35ByhMuw*
---
## *contact me* 
* *[email](mailto:051.SYS113@gmail.com)*
* *[telegram](https://t.me/SYS113/)*
* *[instagram](https://instagram.com/sys113/)*
---
## *thanks*
+ *mohamad moradiyani for negar module logo*
+ *Hamed 169 for optimize negar module source code*
---
## *last word*
*hope this is <ins>negar</ins> useful to you and enjoy it.*

---


CHANGELOG
+ 0.1.0

	 - license
		* gpl v3.0
    
	-  set log file size
		* default is 2
    		* range 1 ... 5
    
    	- set log file name
		* default log.txt
		
	- auto python source code line detection
		* maximum python source code line number support is 999999 
	
	- auto python source code file name detection
		* maximum size of python file name support is 15 character
		
	- auto date derection
	
	- auto time detection
	
	- numbering for each line
		* maximum number to numbering lines support is 9999999
		
	- auto continent detection
		* no need network connection
	
	- auto country detection
		* no need network connection
	
	- auto city capital detection
		* no need network connection
		
	- auto username detection
	
	- auto os detection
	
	- auto os version detection
	
	- auto architecture detection
	
+ 0.2.0

	- create CHANGELOG file
	
	- create TODO file
	
	- change license 
		* from gpl v3.0 to MIT license
	
	- update README.md
	
	- fix bugs reported

+ 0.3.0

	- convert negar source code to be pep8 style ...

	- fix bug write country to log file ...
	
+ 0.4.0

	- add error function
		* write python project exceptions to a file
	
	- rename log function name 
		* from negar.log to negar.text
		
	- rename text argument
		* from negar.text(text='hi') to negar.text(text_log='hi')


+ 0.5.0

	- add 'line' argument in error & text function
		* set 'Flase' and 'True' value for show line python file in log file , default is False ... 

	- add 'time' argument in error & text function
		* set 'Flase' and 'True' value for show line python file in log file , default is False ... 

	- add 'line' argument in error & text function
		* set 'Flase' and 'True' value for show line python file in log file , default is False ... 

+ 0.6.0
	- add 'title' argument in error & text function
		* set str value for show text in log file text , default is not show custom str , default is 'city < country < continent | user name | os > os version > architecture'


negar/countriesWithTheirCapital.py
countries = {
    "kabul": "afghanistan",
    "tirana": "albania",
    "alger": "algeria",
    "fagatogo": "american samoa",
    "andorra la vella": "andorra",
    "luanda": "angola",
    "the valley": "anguilla",
    "null": "united states minor outlying islands",
    "saint john's": "antigua and barbuda",
    "buenos aires": "argentina",
    "yerevan": "armenia",
    "oranjestad": "aruba",
    "canberra": "australia",
    "wien": "austria",
    "baku": "azerbaijan",
    "nassau": "bahamas",
    "al-manama": "bahrain",
    "dhaka": "bangladesh",
    "bridgetown": "barbados",
    "minsk": "belarus",
    "bruxelles [brussel]": "belgium",
    "belmopan": "belize",
    "porto-novo": "benin",
    "hamilton": "bermuda",
    "thimphu": "bhutan",
    "la paz": "bolivia",
    "sarajevo": "bosnia and herzegovina",
    "gaborone": "botswana",
    "brasília": "brazil",
    "bandar seri begawan": "brunei",
    "sofia": "bulgaria",
    "ouagadougou": "burkina faso",
    "bujumbura": "burundi",
    "phnom penh": "cambodia",
    "yaound": "cameroon",
    "ottawa": "canada",
    "praia": "cape verde",
    "george town": "cayman islands",
    "bangui": "central african republic",
    "n'djam": "chad",
    "santiago de chile": "chile",
    "peking": "china",
    "flying fish cove": "christmas island",
    "west island": "cocos (keeling) islands",
    "santaf": "colombia",
    "moroni": "comoros",
    "brazzaville": "congo",
    "avarua": "cook islands",
    "san jos": "costa rica",
    "zagreb": "croatia",
    "la habana": "cuba",
    "nicosia": "cyprus",
    "praha": "czech republic",
    "copenhagen": "denmark",
    "djibouti": "djibouti",
    "roseau": "dominica",
    "santo domingo de guzm": "dominican republic",
    "dili": "east timor",
    "quito": "ecuador",
    "cairo": "egypt",
    "san salvador": "el salvador",
    "london": "united kingdom",
    "malabo": "equatorial guinea",
    "asmara": "eritrea",
    "tallinn": "estonia",
    "addis abeba": "ethiopia",
    "stanley": "falkland islands",
    "tórshavn": "faroe islands",
    "suva": "fiji islands",
    "helsinki [helsingfors]": "finland",
    "paris": "france",
    "cayenne": "french guiana",
    "papeete": "french polynesia",
    "libreville": "gabon",
    "banjul": "gambia",
    "tbilisi": "georgia",
    "berlin": "germany",
    "accra": "ghana",
    "gibraltar": "gibraltar",
    "athenai": "greece",
    "nuuk": "greenland",
    "saint george's": "grenada",
    "basse-terre": "guadeloupe",
    "aga": "guam",
    "ciudad de guatemala": "guatemala",
    "conakry": "guinea",
    "bissau": "guinea-bissau",
    "georgetown": "guyana",
    "port-au-prince": "haiti",
    "citt": "holy see (vatican capital state)",
    "tegucigalpa": "honduras",
    "victoria": "seychelles",
    "budapest": "hungary",
    "reykjav": "iceland",
    "new delhi": "india",
    "jakarta": "indonesia",
    "tehran": "iran",
    "baghdad": "iraq",
    "dublin": "ireland",
    "jerusalem": "israel",
    "roma": "italy",
    "yamoussoukro": "ivory coast",
    "kingston": "norfolk island",
    "tokyo": "japan",
    "amman": "jordan",
    "astana": "kazakhstan",
    "nairobi": "kenya",
    "bairiki": "kiribati",
    "kuwait": "kuwait",
    "bishkek": "kyrgyzstan",
    "vientiane": "laos",
    "riga": "latvia",
    "beirut": "lebanon",
    "maseru": "lesotho",
    "monrovia": "liberia",
    "tripoli": "libyan arab jamahiriya",
    "vaduz": "liechtenstein",
    "vilnius": "lithuania",
    "luxembourg [luxemburg/l": "luxembourg",
    "macao": "macao",
    "skopje": "north macedonia",
    "antananarivo": "madagascar",
    "lilongwe": "malawi",
    "kuala lumpur": "malaysia",
    "male": "maldives",
    "bamako": "mali",
    "valletta": "malta",
    "dalap-uliga-darrit": "marshall islands",
    "fort-de-france": "martinique",
    "nouakchott": "mauritania",
    "port-louis": "mauritius",
    "mamoutzou": "mayotte",
    "ciudad de m": "mexico",
    "palikir": "micronesia, federated states of",
    "chisinau": "moldova",
    "monaco-ville": "monaco",
    "ulan bator": "mongolia",
    "plymouth": "montserrat",
    "rabat": "morocco",
    "maputo": "mozambique",
    "rangoon (yangon)": "myanmar",
    "windhoek": "namibia",
    "yaren": "nauru",
    "kathmandu": "nepal",
    "amsterdam": "netherlands",
    "willemstad": "netherlands antilles",
    "noum": "new caledonia",
    "wellington": "new zealand",
    "managua": "nicaragua",
    "niamey": "niger",
    "abuja": "nigeria",
    "alofi": "niue",
    "pyongyang": "north korea",
    "belfast": "northern ireland",
    "garapan": "northern mariana islands",
    "oslo": "norway",
    "masqat": "oman",
    "islamabad": "pakistan",
    "koror": "palau",
    "gaza": "palestine",
    "ciudad de panam": "panama",
    "port moresby": "papua new guinea",
    "asunci": "paraguay",
    "lima": "peru",
    "manila": "philippines",
    "adamstown": "pitcairn",
    "warszawa": "poland",
    "lisboa": "portugal",
    "san juan": "puerto rico",
    "doha": "qatar",
    "saint-denis": "reunion",
    "bucuresti": "romania",
    "moscow": "russian federation",
    "kigali": "rwanda",
    "jamestown": "saint helena",
    "basseterre": "saint kitts and nevis",
    "castries": "saint lucia",
    "saint-pierre": "saint pierre and miquelon",
    "kingstown": "saint vincent and the grenadines",
    "apia": "samoa",
    "san marino": "san marino",
    "s": "sao tome and principe",
    "riyadh": "saudi arabia",
    "edinburgh": "scotland",
    "dakar": "senegal",
    "freetown": "sierra leone",
    "singapore": "singapore",
    "bratislava": "slovakia",
    "ljubljana": "slovenia",
    "honiara": "solomon islands",
    "mogadishu": "somalia",
    "pretoria": "south africa",
    "seoul": "south korea",
    "juba": "south sudan",
    "madrid": "spain",
    "khartum": "sudan",
    "paramaribo": "suriname",
    "longyearbyen": "svalbard and jan mayen",
    "mbabane": "swaziland",
    "stockholm": "sweden",
    "bern": "switzerland",
    "damascus": "syria",
    "dushanbe": "tajikistan",
    "dodoma": "tanzania",
    "bangkok": "thailand",
    "lom": "togo",
    "fakaofo": "tokelau",
    "nuku'alofa": "tonga",
    "port-of-spain": "trinidad and tobago",
    "tunis": "tunisia",
    "ankara": "turkey",
    "ashgabat": "turkmenistan",
    "cockburn town": "turks and caicos islands",
    "funafuti": "tuvalu",
    "kampala": "uganda",
    "kyiv": "ukraine",
    "abu dhabi": "united arab emirates",
    "washington": "united states",
    "montevideo": "uruguay",
    "toskent": "uzbekistan",
    "port-vila": "vanuatu",
    "caracas": "venezuela",
    "hanoi": "vietnam",
    "road town": "virgin islands, british",
    "charlotte amalie": "virgin islands, u.s.",
    "cardiff": "wales",
    "mata-utu": "wallis and futuna",
    "el-aai": "western sahara",
    "sanaa": "yemen",
    "beograd": "yugoslavia",
    "lusaka": "zambia",
    "harare": "zimbabwe"
}


negar/negar.py
# Copyright SYS113 2019. MIT license , see README.md file.

# import libraries 
from re import search
from traceback import format_exc
from tzlocal import get_localzone
from datetime import datetime
from platform import system, release, machine
from getpass import getuser
from os.path import isfile
from inspect import getframeinfo, stack
from negar.countriesWithTheirCapital import countries


# helper function for country capital
def get_country(_city):
	data = countries
	if _city in data:
		return data[_city]
	else:
		return 'unknown'


# helper function for negar module errors printing
def err_temp_func(file_, line, problem):
	error_template = 'negar module - error | python file : {} | line : {} | problem : {}'
	return error_template.format(file_, line, problem)


# helper function for justify text center with fixed length
def justify_text(text_, length):
	return '{}{}{}'.format(int((length - len(str(text_))) / 2) * ' ', text_, length * ' ')[:length]


# helper function for create header row
def header_row(header_specs):
	sep = '—' * (len(header_specs) - 4)
	out = '\n  .{1}.\n {0}\n  |{1}|\n'.format(header_specs, sep)
	return out


# helper function for create each log row
def log_row(row_num, log_date, log_time, row_text, row_log_size, row_file_name, row_type, row_line_num):
	row_num = justify_text(row_num, 7)
	row_type = justify_text(row_type, 8)
	out = '  |{num}| {date} | {time} | {text}{pad}|{file}|{type}|{line}|\n'.format(
		num=row_num, date=log_date, time=log_time, text=row_text, file=row_file_name, type=row_type,
		line=row_line_num, pad=' ' * (row_log_size - (len(row_text) + 1)))
	out += '  |{}|\n'.format('—' * (len(out) - 5))
	return out


# create text log function ...
def text(text_log=None, save=None, size=None , line=None , date=None , time=None , title=None):

	# find (filename or line) python file ...
	x = stack()[1]
	x = x[0]
	get_log_file_python_file_name_or_line = getframeinfo(x)

	# ----------------------------------------------------------------------------------------------------------------
	'''                                                    
	'find line in python file' variable is 'line_python_file'
	'find python file name' variable is 'python_file'
	'save log in a file' variable is 'log_file'
	'log text' variable is 'log_text'
	'set log file size' variable is 'log_file_size'
	'''
	# ----------------------------------------------------------------------------------------------------------------

	# set Flase and True value for show line python file in log file ...
	if (line is None) or (line is False):
		line_python_file = '-'
	elif line is True:
		line_python_file = str(get_log_file_python_file_name_or_line.lineno)
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"show" value is not boolean ...'))
		return

	# set value for find python file name ...
	python_file_name = str(get_log_file_python_file_name_or_line.filename.split('/')[-1])

	if python_file_name in ['<stdin>', '<input>']:
		python_file = 'interpreter'
	elif len(python_file_name) < 18:
		python_file = python_file_name
	else:
		print(err_temp_func(python_file_name, line_python_file, 'maximum size of python file name is 15 character ...'))
		return

	# set value for save log in a file ...
	if isinstance(save, str):
		log_file = save
	elif save is None:
		log_file = 'log.txt'
	else:
		print(err_temp_func(python_file_name, line_python_file, '"save" type is not str ...'))
		return

	# set value for log text ...
	if text_log is None:
		print(err_temp_func(python_file_name, line_python_file, '"text" value is empty ...'))
		return
	elif not isinstance(text_log, str):
		print(err_temp_func(python_file_name, line_python_file, '"text" type is not str ...'))
		return
	elif text_log == '':
		print(err_temp_func(python_file_name, line_python_file, '"text" value is empty ...'))
		return
	else:
		log_text = text_log

	# set value for log file size ...
	if size is None:
		log_size = 3
	elif not isinstance(size, int):
		print(err_temp_func(python_file_name, line_python_file, '"size" type is not str ...'))
		return
	elif size > 5:
		print(err_temp_func(python_file_name, line_python_file, '"size" value range is not (1 ... 5) ...'))
		return
	elif size == 0:
		print(err_temp_func(python_file_name, line_python_file, '"size" value is not in range (1 ... 5) ...'))
		return
	else:
		log_size = size

	# ----------------------------------------------------------------------------------------------------------------
	'''
	'write python file name in log file' variable is 'justified_python_file'
	'write line of python file in log file' variable is 'justified_line_python_file'
	'write log text in log file' variable is 'log_file_text'
	'log file name' variable is 'log_file_name'
	'log file size' variable is 'log_file_size'
	'''
	# ----------------------------------------------------------------------------------------------------------------

	# check character size ...
	if (log_size == 1) and (len(log_text) > 69):
		print(err_temp_func(python_file_name, line_python_file, '"size = 1" maximum 69 character support ...'))
		return
	elif (log_size == 2) and (len(log_text) > 93):
		print(err_temp_func(python_file_name, line_python_file, '"size = 2" maximum 93 character support ...'))
		return
	elif (log_size == 3) and (len(log_text) > 131):
		print(err_temp_func(python_file_name, line_python_file, '"size = 3" maximum 131 character support ...'))
		return
	elif (log_size == 4) and (len(log_text) > 199):
		print(err_temp_func(python_file_name, line_python_file, '"size = 4" maximum 199 character support ...'))
		return
	elif (log_size == 5) and (len(log_text) > 397):
		print(err_temp_func(python_file_name, line_python_file, '"size = 5" maximum 397 character support ...'))
		return

	# set value for write python file name in log file ...
	justified_python_file = justify_text(python_file, 18)

	# set value for write line of python file in log file ...
	if len(line_python_file) <= 6:
		justified_line_python_file = justify_text(line_python_file, 6)
	else:
		print(err_temp_func(python_file_name, line_python_file, 'maximum python line number support is 999999 ...'))
		return

	# set value for log file name ...
	log_file_name = log_file

	# set value for log file size ...
	log_file_size = round(((6 - 0.3) / 100) * ((5 - [0.1, 2.9, 3.9, 4.4, 4.7][log_size - 1]) * 20), 1)
	# ----------------------------------------------------------------------------------------------------------------

	# get continent ...
	continent = str(get_localzone()).lower().split('/')[0]

	# get city ...
	city = str(get_localzone()).lower().split('/')[1]

	# get country ...
	country = str(get_country(city))

	# get username ...
	username = str(getuser())

	# get os ...
	os = str(system().lower())

	# get version ...
	version = str(release().lower())

	# get architecture
	architecture = str(machine().lower())

	# log type
	log_type = 'text'

	# ----------------------------------------------------------------------------------------------------------------
	# set Flase and True value for time in log file ...
	if (time is None) or (time is True):
		time = str(datetime.now()).split(' ')[1].split('.')[0]
	elif time is False:
		time = '   -    '
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"time" value is not boolean ...'))
		return
	# ----------------------------------------------------------------------------------------------------------------
	# set Flase and True value for show date in log file ...
	if (date is None) or (date is True):
		date = str(datetime.now()).split(' ')[0]
	elif date is False:
		date = '    -     '
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"date" value is not boolean ...'))
		return
	# ----------------------------------------------------------------------------------------------------------------

	# show city , country , continent , username , os , version , architecture in log file ...
	sp_center = '%s < %s < %s | %s | %s > %s > %s' % (
		city, country, continent, username, os, version, architecture)
	spec_center = justify_text(sp_center, 2 * int(len(sp_center) / log_file_size) + len(sp_center))
	T = (len(spec_center)-len(sp_center))/2

	if title is not None:
		if len(title) > len(spec_center):
			print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), 'maximum \'title\' size is '+str(len(spec_center))+' ...'))
			return

	if (title is None) or (title is False):
		specifications = ' |  num  |    date    |   time   |' + spec_center + '|       file       |  type  | line | '
	else:
		specifications = ' |  num  |    date    |   time   |' + title.center(len(spec_center)) + '|       file       |  type  | line | '

	# if is not log file ...
	if not isfile(log_file_name):
		# create log file ...
		with open(log_file_name, 'w') as log_file:
			# write ' ___ ' to log file ...
			log_file.write(header_row(specifications))

			# write 'error & text' to log file ...
			log_file.write(
				log_row(1, date, time, text_log, len(spec_center), justified_python_file, log_type,
						justified_line_python_file))

	# if is log file ...
	elif isfile(log_file_name):
		with open(log_file_name, 'r') as f:
			if f.read().splitlines()[-1] != '  |' + '—' * (len(specifications) - 4) + '|':
				print(err_temp_func(python_file_name, line_python_file,
									"previously defined log file size , can't be resized ..."))
				return

		# find line number ...
		with open(log_file_name, 'r') as f:
			line_number = f.read().splitlines()[-2].split('|')[1]
			line_number = line_number.replace(' ', '')
			line_number = int(line_number) + 1

		# ckeck log file line number ...
		if len(str(line_number)) > 7:
			print(err_temp_func(python_file_name, line_python_file,
								"'size = 5' maximum line number support is 9999999 ..."))
			return

		# set value for log file line number ...
		if len(str(line_number)) <= 7:
			log_file_number = justify_text(line_number, 7)
		else:
			print(err_temp_func(python_file_name, line_python_file,
								'maximum number to numbering lines support is 9999999 ...'))
			return

		# open log file ...
		with open(log_file_name, 'a') as log_file:

			# add new 'error & text' to log file ...
			log_file.write(log_row(log_file_number, date, time, text_log, len(spec_center), justified_python_file,
								   log_type, justified_line_python_file))


# create error log function ...
def error(save=None, size=None, line=None , time=None , date=None , title=None):

	# find (filename or line) python file ...
	x = stack()[1]
	x = x[0]
	get_log_file_python_file_name_or_line = getframeinfo(x)

	# write exception to error_log variable ...
	error_log = format_exc()
	if len(error_log.splitlines()) < 2:
		return

	# error log ...
	error_text = str(error_log.splitlines()[-1])
	a = [*filter(lambda x: x.find('File') > -1, error_log.splitlines())]

	# file name ...
	python_file_name = search('File "(.*)", line (.*), in (.*)', a[-1]).groups()
	for i in a[:-1][::-1]:
		x = search('File "(.*)", line (.*), in (.*)', i).groups()
		if x and python_file_name[2] == x[2]:
			python_file_name = x
		else:
			break

	# set Flase and True value for show line python file in log file ...
	if (line is None) or (line is True):
		line_python_file = str(get_log_file_python_file_name_or_line.lineno)
	elif line is False:
		line_python_file = '-'
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"line" value is not boolean ...'))
		return

	# name python file ...
	python_file_name = str(python_file_name[0]).split('/')[-1]

	# ----------------------------------------------------------------------------------------------------------------
	'''                                                    
	'find line in python file' variable is 'line_python_file'
	'find python file name' variable is 'python_file'
	'save log in a file' variable is 'log_file'
	'log text' variable is 'log_text'
	'set log file size' variable is 'log_file_size'
	'''
	# ----------------------------------------------------------------------------------------------------------------

	if python_file_name in ['<stdin>', '<input>']:
		python_file = 'interpreter'
	elif len(python_file_name) < 18:
		python_file = python_file_name
	else:
		print(err_temp_func(python_file_name, line_python_file, 'maximum size of python file name is 15 character ...'))
		return

	# set value for save log in a file ...
	if isinstance(save, str):
		log_file = save
	elif save is None:
		log_file = 'log.txt'
	else:
		print(err_temp_func(python_file_name, line_python_file, '"save" type is not str ...'))
		return

	# set value for log file size ...
	if size is None:
		log_size = 3
	elif not isinstance(size, int):
		print(err_temp_func(python_file_name, line_python_file, '"size" type is not str ...'))
		return
	elif size > 5:
		print(err_temp_func(python_file_name, line_python_file, '"size" value range is not (1 ... 5) ...'))
		return
	elif size == 0:
		print(err_temp_func(python_file_name, line_python_file, '"size" value is not in range (1 ... 5) ...'))
		return
	else:
		log_size = size

	# ----------------------------------------------------------------------------------------------------------------
	'''
	'write python file name in log file' variable is 'justified_python_file'
	'write line of python file in log file' variable is 'justified_line_python_file'
	'write log text in log file' variable is 'log_file_text'
	'log file name' variable is 'log_file_name'
	'log file size' variable is 'log_file_size'
	'''
	# ----------------------------------------------------------------------------------------------------------------

	# check character size ...
	if (log_size == 1) and (len(error_text) > 69):
		print(err_temp_func(python_file_name, line_python_file, '"size = 1" maximum 69 character support ...'))
		return
	elif (log_size == 2) and (len(error_text) > 93):
		print(err_temp_func(python_file_name, line_python_file, '"size = 2" maximum 93 character support ...'))
		return
	elif (log_size == 3) and (len(error_text) > 131):
		print(err_temp_func(python_file_name, line_python_file, '"size = 3" maximum 131 character support ...'))
		return
	elif (log_size == 4) and (len(error_text) > 199):
		print(err_temp_func(python_file_name, line_python_file, '"size = 4" maximum 199 character support ...'))
		return
	elif (log_size == 5) and (len(error_text) > 397):
		print(err_temp_func(python_file_name, line_python_file, '"size = 5" maximum 397 character support ...'))
		return

	# set value for write python file name in log file ...
	justified_python_file = justify_text(python_file, 18)

	# set value for write line of python file in log file ...
	if len(line_python_file) <= 6:
		justified_line_python_file = justify_text(line_python_file, 6)
	else:
		print(err_temp_func(python_file_name, line_python_file, 'maximum python line number support is 999999 ...'))
		return

	# set value for log file name ...
	log_file_name = log_file

	# set value for log file size ...
	log_file_size = round(((6 - 0.3) / 100) * ((5 - [0.1, 2.9, 3.9, 4.4, 4.7][log_size - 1]) * 20), 1)
	# ----------------------------------------------------------------------------------------------------------------
	# get continent ...
	continent = str(get_localzone()).lower().split('/')[0]

	# get city ...
	city = str(get_localzone()).lower().split('/')[1]

	# get country ...
	country = str(get_country(city))

	# get username ...
	username = str(getuser())

	# get os ...
	os = str(system().lower())

	# get version ...
	version = str(release().lower())

	# get architecture
	architecture = str(machine().lower())

	# log type
	log_type = ' error'
	# ----------------------------------------------------------------------------------------------------------------
	# set Flase and True value for time in log file ...
	if (time is None) or (time is True):
		time = str(datetime.now()).split(' ')[1].split('.')[0]
	elif time is False:
		time = '   -    '
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"time" value is not boolean ...'))
		return
	# ----------------------------------------------------------------------------------------------------------------
	# set Flase and True value for show date in log file ...
	if (date is None) or (date is True):
		date = str(datetime.now()).split(' ')[0]
	elif date is False:
		date = '    -     '
	else:
		print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), '"date" value is not boolean ...'))
		return
	# ----------------------------------------------------------------------------------------------------------------
	'''
	# show city , country , continent , username , os , version , architecture in log file ...
	spec_center = '%s < %s < %s | %s | %s > %s > %s' % (
		city, country, continent, username, os, version, architecture)
	spec_center = justify_text(spec_center, 2 * int(len(spec_center) / log_file_size) + len(spec_center))
	specifications = ' |  num  |    date    |   time   |' + spec_center + '|       file       |  type  | line | '
	'''

	# show city , country , continent , username , os , version , architecture in log file ...
	sp_center = '%s < %s < %s | %s | %s > %s > %s' % (
		city, country, continent, username, os, version, architecture)
	spec_center = justify_text(sp_center, 2 * int(len(sp_center) / log_file_size) + len(sp_center))
	T = (len(spec_center)-len(sp_center))/2
	
	if title is not None:
		if len(title) > len(spec_center):
			print(err_temp_func(str(get_log_file_python_file_name_or_line.filename.split('/')[-1]), str(get_log_file_python_file_name_or_line.lineno), 'maximum \'title\' size is '+str(len(spec_center))+' ...'))
			return

	if (title is None) or (title is False):
		specifications = ' |  num  |    date    |   time   |' + spec_center + '|       file       |  type  | line | '
	else:
		specifications = ' |  num  |    date    |   time   |' + title.center(len(spec_center)) + '|       file       |  type  | line | '


	# if is not log file ...
	if not isfile(log_file_name):

		# create log file ...
		with open(log_file_name, 'w') as log_file:

			# write ' ___ ' to log file ...
			log_file.write(header_row(specifications))

			# write 'error & text' to log file ...
			log_file.write(
				log_row(1, date, time, error_text, len(spec_center), justified_python_file, log_type,
						justified_line_python_file))

	# if is log file ...
	elif isfile(log_file_name):
		with open(log_file_name, 'r') as f:
			if f.read().splitlines()[-1] != '  |' + '—' * (len(specifications) - 4) + '|':
				print(err_temp_func(python_file_name, line_python_file,
									"previously defined log file size , can't be resized ..."))
				return

		# find line number ...
		with open(log_file_name, 'r') as f:
			line_number = f.read().splitlines()[-2].split('|')[1]
			line_number = line_number.replace(' ', '')
			line_number = int(line_number) + 1

		# ckeck log file line number ...
		if len(str(line_number)) > 7:
			print(err_temp_func(python_file_name, line_python_file,
								"'size = 5' maximum line number support is 9999999 ..."))
			return

		# set value for log file line number ...
		if len(str(line_number)) <= 7:
			log_file_number = justify_text(line_number, 7)
		else:
			print(err_temp_func(python_file_name, line_python_file,
								'maximum number to numbering lines support is 9999999 ...'))
			return

		# open log file ...
		with open(log_file_name, 'a') as log_file:
			# add new 'error & text' to log file ...
			log_file.write(log_row(log_file_number, date, time, error_text, len(spec_center), justified_python_file,
								   log_type, justified_line_python_file))
